kernel/comps/device/src/id.rs

```diff
@@ -0,0 +1,284 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::collections::btree_set::BTreeSet;
+use core::{fmt::Debug, ops::Range};
+
+use ostd::sync::RwLock;
+
+/// The number of bits used to represent the major device number.
+pub const MAJOR_BITS: usize = 12;
+
+/// The number of bits used to represent the minor device number.
+pub const MINOR_BITS: usize = 20;
+
+/// The type of a device.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DeviceType {
+    /// Block devices, which transfer data in fixed-size blocks.
+    Block,
+    /// Character devices, which transfer data as a stream of bytes.
+    Char,
+    /// Other device types that don't fit into the block or character categories.
+    Other,
+}
+
+/// A device identifier consisting of a major and minor number.
+///
+/// Device IDs are used to uniquely identify devices in the system. The major number
+/// identifies the device driver, while the minor number identifies a specific device
+/// instance controlled by that driver.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct DeviceId {
+    major: u32,
+    minor: u32,
+}
+
+impl DeviceId {
+    fn new(major: u32, minor: u32) -> Self {
```

Comment:
This constructor should be made `pub` because it gives a false sense of safety. It appears to be saying that we want `DeviceId` to be created ony through `DeviceIdAllocator` so I need to hide the constructor method. But you have provided the more powerful `from_encoded_u64`, which could create a device ID from arbitrary raw values. So hiding the `new` constructor is meaningless and appears to be double standard.